lib/exprengine.cpp

```diff
@@ -977,21 +977,41 @@ static bool isNonOverlapping(ExprEngine::ValuePtr v1, ExprEngine::ValuePtr v2)
 
 static bool conditionAlwaysFalse(ExprEngine::ValuePtr condValue, ExprEngine::DataBase *dataBase)
 {
-    if (auto v = std::dynamic_pointer_cast<ExprEngine::IntRange>(condValue))
-        return v->hasValue(0);
-    if (auto v = std::dynamic_pointer_cast<ExprEngine::FloatRange>(condValue))
-        return v->hasValue(0.0);
+    if (auto v = std::dynamic_pointer_cast<ExprEngine::IntRange>(condValue)) {
+        if (v->hasValue(0))
```

Comment:
[`hasValue`](https://github.com/danmar/cppcheck/tree/9a511b34c1937487f33621e4392e0277a8e13416/lib/exprengine.h#L159) simply compare two integers. I don't want to drag `Token` value there, because this method could be reused in other contexts, when we don't have a specific token instance. I also don't want to validate a token outside of this function, because it complicates the interface.

Reply:
I feel a bit scared about this anyway.. there could be constraints in the execution path and you ignore those:

    void foo(int x) { //x:$1
        int zero = 15;  // zero:$2
        if (x > 100 || zero > 100) {
            // there is a constraint in this execution path:  $1>100 || $2>100  => true
        }
    }

I cannot give a concrete example where hasValue would fail. but it feels unfortunate to ignore constraints.

Well if you would rename it to `hasKnownValue` then it would reuse the terminology from ValueFlow a bit at least.